### PR TITLE
Make LIFT_FUNCTION more modular

### PR DESCRIPTION
**Pull request by andralex**
Per the discussion at https://twitter.com/incomputable/status/1623722860610039809

### CODE REVIEW

File: include/lift.hpp [hunk 1]
```diff
@@ -17,16 +17,17 @@
...
-        {                               \
-          return __VA_ARGS__;           \
-        }
-
-#define LIFT_FWD(x) std::forward<decltype(x)>(x)
+#define LIFT_FUNCTION(f)                                                  \
+    LIFT_ba7b8453f262e429575e23dcb2192b33(                                \
+        a_ba7b8453f262e429575e23dcb2192b33,                               \
+        (f)(::std::forward<decltype(a_ba7b8453f262e429575e23dcb2192b33)>( \
+            a_ba7b8453f262e429575e23dcb2192b33)...))
```
Comment (andralex) [inline include/lift.hpp:24]: Actually `LIFT(f<int, int>)` should work, so...

```suggestion
#define LIFT_FUNCTION(...)                                                  \
    LIFT_ba7b8453f262e429575e23dcb2192b33(                                \
        a_ba7b8453f262e429575e23dcb2192b33,                               \
        (__VA_ARGS__)(::std::forward<decltype(a_ba7b8453f262e429575e23dcb2192b33)>( \
            a_ba7b8453f262e429575e23dcb2192b33)...))
```

**Comment by rollbear** [comment]
Thanks Andrei,

I'm a little bit confused. Is there a reason for not going with the simpler:

```diff
diff --git a/include/lift.hpp b/include/lift.hpp
index 39cfdc6..fa2a570 100644
--- a/include/lift.hpp
+++ b/include/lift.hpp
@@ -26,7 +26,7 @@
 
 #define LIFT_FWD(x) std::forward<decltype(x)>(x)
 
-#define LIFT_FUNCTION(lift_func) [](auto&& ... p) LIFT_THRICE(lift_func(LIFT_FWD(p)...))
+#define LIFT_FUNCTION(...) [](auto&& ... p) LIFT_THRICE(__VA_ARGS__(LIFT_FWD(p)...))
 #ifndef LIFT
 #define LIFT LIFT_FUNCTION
 #endif
```

Also, an added test case or two would've been nice.

**Comment by andralex** [comment]
@rollbear you mean why use a variadic macro for `LIFT_FUNCTION`? Consider a call such as `LIFT_FUNCTION(Widget<int, int>::fun)`. From the compiler's point of view, `Widget<int, int>::fun` is a single entity. But the preprocessor does not "pair" angle brackets correctly so it thinks you are invoking `LIFT_FUNCTION` with two arguments, `Widget<int` and `int>::fun`. If the macro is defined with one parameter, that'll be a preprocessing error. If you use `...` instead, the macro will accept everything and expand correctly.

**Comment by andralex** [comment]
Anyway, the gist of this PR is the intentional obfuscation for achieving some sort of modularity for macro names. You don't want to pollute the perilous macro namespace with names you don't plan to export (such as `LIFT_THRICE` of `LIFT_FWD`.

**Comment by rollbear** [comment]
Thanks Andrei!

**Pull request merged**